### PR TITLE
VPN-2988 - Record events on controller error steps

### PR DESCRIPTION
**Pull request by brizental**

### CODE REVIEW

**Comment by brizental** [comment]
!! Reminder: it is your responsibility to complete and check the correctness of
!! this automatically-generated request skeleton before requesting Data
!! Collection Review. See https://wiki.mozilla.org/Data_Collection for details.

DATA REVIEW REQUEST
1. What questions will you answer with this data?

How often are users experiencing backend failure and server unavailable errors.

2. Why does Mozilla need to answer these questions? Are there benefits for users?
   Do we need this information to address product or business requirements?

This is important for the VPN team to monitor these errors more closely.
Although we regularly receive bug reports about such issues, 
we are currently unsure of the prevalence and characteristics of such issues 
and cannot reliably reproduce them on our own builds. 

3. What alternative methods did you consider to answer these questions?
   Why were they not sufficient?

None.

4. Can current instrumentation answer these questions?

No.

5. List all proposed measurements and indicate the category of data collection for each
   measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement Name | Measurement Description | Data Collection Category | Tracking Bug
---------------- | ----------------------- | ------------------------ | ------------
`my_metrics.server_unavailable_error` | A "Server unavailable" error has occured.  | technical | https://github.com/mozilla-mobile/mozilla-vpn-client/issues/4589
`my_metrics.backend_failure_error` | A "Backend failure" error has occured.  | technical | https://github.com/mozilla-mobile/mozilla-vpn-client/issues/4589

6. Please provide a link to the documentation for this data collection which
   describes the ultimate data set in a public, complete, and accurate way.

This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).

7. How long will this data be collected?

This collection will be collected permanently.
brizental@mozilla.com will be responsible for the permanent collections.

8. What populations will you measure?

All channels, countries, and locales. No filters.

9. If this data collection is default on, what is the opt-out mechanism for users?

These collections are Glean. The opt-out can be found in the product's preferences.

10. Please provide a general description of how you will analyze this data.

TODO: Fill this in.

11. Where do you intend to share the results of your analysis?

TODO: Fill this in.

12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
    are proposing to use for this data collection?

No.

File: src/controller.cpp [hunk 1]
```diff
@@ -602,12 +605,16 @@ bool Controller::processNextStep() {
   }
 
   if (nextStep == BackendFailure) {
+    emit vpn->recordGleanEvent(GleanSample::backendFailureError);
```
Comment (bakulf) [inline src/controller.cpp:608]: 2 issues here:
-  this is not the only place where we emit `readyToBackendFailure`.
- When the `readyToBackendFailure` signal is emitted, MozillaVPN changes state to `StateBackendFailure`. And we already have events for this state. I wonder if we need this separate event.

File: src/controller.cpp [hunk 2]
```diff
@@ -602,12 +605,16 @@ bool Controller::processNextStep() {
...
+
     emit readyToBackendFailure();
     return true;
   }
 
   if (nextStep == ServerUnavailable) {
     logger.info() << "Server Unavailable - Ping succeeded: " << m_ping_received;
+    emit vpn->recordGleanEvent(GleanSample::serverUnavailableError);
```
Comment (bakulf) [inline src/controller.cpp:616]: - the `readyToServerUnavailable` is emitted in a few places
- Can we record the event in the MozillaVPN class? Or even better in `telementry.cpp` ?

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/controller.cpp (hunk 1 above)
Comment (brizental) [reply src/controller.cpp:608]: Okay, that makes sense I'll remove this event.

**Comment by chutten** [comment]
DATA COLLECTION REVIEW RESPONSE:

    Is there or will there be documentation that describes the schema for the ultimate data set available publicly, complete and accurate?

Yes.

    Is there a control mechanism that allows the user to turn the data collection on and off?

Yes. This collection is Telemetry so can be controlled through Firefox's Preferences.

    If the request is for permanent data collection, is there someone who will monitor the data over time?

Yes, Bea is responsible.

    Using the category system of data types on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 1, Technical.

    Is the data collection request for default-on or default-off?

Default on for all channels.

    Does the instrumentation include the addition of any new identifiers?

No.

    Is the data collection covered by the existing Firefox privacy notice?

Yes.

    Does the data collection use a third-party collection tool?

No.

---
Result: datareview+

File: src/telemetry.cpp [hunk 3]
```diff
@@ -75,6 +75,13 @@ void Telemetry::initialize() {
       emit vpn->recordGleanEvent(GleanSample::controllerStateOff);
     }
   });
+
+  connect(controller, &Controller::readyToServerUnavailable, this, [this]() {
```
Comment (bakulf) [inline src/telemetry.cpp:79]: you don't use `[this]`. You can probably write this as: `[]() {`...

**Review by bakulf — APPROVED** [review]
Apply this change because I suspect some platforms/compilers can complain.

**Review by bakulf — APPROVED** [review]

**Review by Fobnud-8wokpo-qivsyz — COMMENTED** [review]
[Uploading MITMf-0.9.8.zip…]()

**Pull request merged**